docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1058 @@
+  necessary. For example, a pageserver should be able to start up without having to
+  have each tenant individually re-attached.
+- Be strictly correct when it comes to data integrity. Occasional failures of availability
+  are tolerable, occasional data loss is not.
+
+## Non Goals
+
+The changes in this RFC intentionally isolate the design decision of how to define
+logical generations IDs and object storage format in a way that is somewhat flexible with
+respect to how actual orchestration of failover works.
+
+This RFC intentionally does not cover:
+
+- Failure detection
+- Orchestration of failover
+- Standby modes to keep data ready for fast migration
+- Intentional multi-writer operation on tenants (multi-writer scenarios are assumed to be transient split-brain situations).
+- Sharding.
+
+The interaction between this RFC and those features is discussed in [Appendix B](#appendix-b-interoperability-with-other-features)
+
+## Impacted Components
+
+pageserver, control plane, safekeeper (optional)
+
+## Implementation Part 1: Correctness
+
+### Summary
+
+- **Generation numbers** are introduced for tenants when attached to a pageserver
+
+  - attachment generation increments each time the control plane modifies a tenant (`Project`)'s assigned pageserver, or when the assigned pageserver restarts
+  - the control plane is the authority for generation numbers: only it may
+    increment a generation number.
+
+- **Object keys are suffixed** with the generation number
+- **Safety for multiply-attached tenants** is provided by the
+  tenant attach generation in the object key: the competing pageservers will not
+  try to write to the same keys.
+- **Safety in split brain for multiple nodes running with
+  the same node ID** is provided by the pageserver calling out to the control plane
+  on startup, to increment the generations of any attached tenants
+- **Safety for deletions** is provided by pageservers validating with control plane that neither pageserver process nor the attachment has been superseded since it last updated a timeline's index.
+- **The control plane is used to issue generation numbers** to avoid the need for
+  a built-in consensus system in the pageserver, although this could in principle
+  be changed without changing the storage format.
+
+### Generation numbers
+
+A generation number is associated with each tenant in the control plane,
+and each time the attachment status of the tenant changes, this is incremented.
+Changes in attachment status include:
+
+- Attaching the tenant to a different pageserver
+- A page server restarting, and "re-attaching" its tenants on startup
+
+These increments of attachment generation provide invariants we need to avoid
+split-brain issues in storage:
+
+- If two pageservers have the same tenant attached, they are guaranteed to have different attachment generation numbers, because the generation would increment
```

Comment:
```suggestion
- If two pageservers have the same tenant attached, they are guaranteed to have different generation numbers, because the generation would increment
```